src/Language/PureScript/Docs/RenderedCode/RenderType.hs

```diff
@@ -156,6 +158,32 @@ forall_ = mkPattern match
 renderType :: Type a -> RenderedCode
 renderType = renderType' . convertPrettyPrintType maxBound
 
+renderTypeWithRole :: [Role] -> Type a -> RenderedCode
+renderTypeWithRole roleList = renderType' . insertRoles . convertPrettyPrintType maxBound
+  where
+  insertRoles :: PrettyPrintType -> PrettyPrintType
+  insertRoles = snd . addRole roleList
+
+  -- `data Foo first second = Foo` will produce
+  -- ```
+  -- PPTypeApp
+  --  (PPTypeApp (PPTypeConstructor fooName) (PPTypeVar "first" Nothing))
+  --  (PPTypeVar "second" Nothing)
+  -- ```
+  -- So, we recurse down the left side of `TypeApp` first before
+  -- recursing down the right side.
+  addRole :: [Role] -> PrettyPrintType -> ([Role], PrettyPrintType)
+  addRole [] pp = ([], pp)
+  addRole roles@(x:xs) pp = case pp of
+    PPTypeApp f a ->
+      let
+        (remainingRoles1, f') = addRole roles f
+        (remainingRoles2, a') = addRole remainingRoles1 a
+      in (remainingRoles2, PPTypeApp f' a')
+    PPTypeVar t Nothing ->
+      (xs, PPTypeVar t (Just $ displayRole x))
+    other -> (roles, other)
```

Comment:
AFAICT, this function isn't stack-safe. However, I don't think that issue would ever be triggered.

Reply:
This has been fixed in latest commit.